test/unit/logic/delegate.js

```diff
@@ -0,0 +1,535 @@
+				u_isDelegate: 1,
+				isDelegate: 0,
+				username: null,
+				u_username: trs.asset.delegate.username
+			}, dummyCb);
+
+			delegate.applyUnconfirmed(trs, sender, dummyCb);
+			accountsMock.setAccountAndGet.verify();
+
+			done();
+		});
+	});
+
+	describe.skip('undoUnconfirmed', function () {
+
+		it('should update username to null if account did not have a username before', function (done) {
+			delete sender.username;
+			sender.nameexist = 0;
+
+			accountsMock.setAccountAndGet.once().withExactArgs({
+				address: sender.address,
+				u_isDelegate: 0,
+				isDelegate: 0,
+				username: null,
+				u_username: null
+			}, dummyCb);
+
+			delegate.undoUnconfirmed(trs, sender, dummyCb);
+			accountsMock.setAccountAndGet.verify();
+
+			done();
+		});
+
+		it('should call accounts.setAccountAndGet module with correct parameters', function (done) {
+			accountsMock.setAccountAndGet.once().withExactArgs({
+				address: sender.address,
+				u_isDelegate: 0,
+				isDelegate: 0,
+				username: null,
+				u_username: null
+			}, dummyCb);
+
+			delegate.undoUnconfirmed(trs, sender, dummyCb);
+			accountsMock.setAccountAndGet.verify();
+
+			done();
+		});
+	});
+
+	describe('objectNormalize', function () {
+
+		var schemaDynamicTest = new SchemaDynamicTest({
+			testStyle: SchemaDynamicTest.TEST_STYLE.THROWABLE,
+			customPropertyAssertion: function (input, expectedType, property, err) {
+				expect(err).to.equal('Failed to validate delegate schema: Expected type ' + expectedType + ' but found type ' + input.expectation);
+			}
+		});
+
+		after(function () {
+			describe('schema dynamic tests', function () {
+				schemaDynamicTest.schema.shouldFailAgainst.nonObject.property(delegate.objectNormalize, trs, 'asset.delegate');
+
+				schemaDynamicTest.schema.shouldFailAgainst.nonString.property(delegate.objectNormalize, trs, 'asset.delegate.username');
+			});
+		});
+
+		it('should use the correct format to validate against', function () {
+			var library = Delegate.__get__('library');
+			var schemaSpy = sinon.spy(library.schema, 'validate');
+			delegate.objectNormalize(trs);
+			expect(schemaSpy.calledOnce).to.equal(true);
+			expect(schemaSpy.calledWithExactly(trs.asset.delegate, Delegate.prototype.schema)).to.equal(true);
+			schemaSpy.restore();
+		});
+
+		it('should throw error when asset schema is invalid', function () {
+			trs.asset.delegate.username = '';
+
+			expect(function () {
+				delegate.objectNormalize(trs);
+			}).to.throw('Failed to validate delegate schema: Object didn\'t pass validation for format username: ');
+		});
+
+		it('should return transaction when asset is valid', function () {
+			expect(delegate.objectNormalize(trs)).to.eql(trs);
+		});
+	});
+
+	describe.skip('dbRead', function () {
+
+		it('should return null when username is not set ', function () {
+			delete rawTrs.d_username;
+
+			expect(delegate.dbRead(rawTrs)).to.eql(null);
+		});
+
+		it('should be okay for valid input', function () {
```

Comment:
what does it mean that it is ok - please test the results returned by callback when result contains `d_username` field

Reply:
I will update the test description, the test is deeply checking the result returned from the function like this:
```
			expect(delegate.dbSave(trs)).to.eql({
				table: 'delegates',
				fields: [
					'tx_id',
					'name',
					'pk',
					'address'
				],
				values: {
					tx_id: trs.id,
					name: trs.asset.delegate.username,
					pk: Buffer.from(trs.senderPublicKey, 'hex'),
					address: trs.senderId
				}
			});
```

Reply:
It checks for all asset properties, the test description is ambiguous, I will update it.